### DISCUSSION TITLE
Add a minified ESM version to the package

### DISCUSSION THREAD

**Discussion by jonkoops** [alpinejs/alpine#4181] [3. Feature ideas]
I quite like using Alpine.js as an ECMAScript module directly in the browser, and it would be really nice if there was already a minified version of this distribution that can be used directly. For example, some of the code I would write now is:

#### index.html
```html
<!doctype html>
<html>
  <head>
    <script type="importmap">
      {
        "imports": {
          "alpinejs": "/node_modules/alpinejs/dist/module.esm.js"
        }
      }
    </script>
    <link rel="modulepreload" href="/node_modules/alpinejs/dist/module.esm.js">
    <script type="module" src="counter.js"></script>
  <body>
    <div x-data="counter">
      <button type="button" @click="increment()">-</button>
      <span x-text="count"></span>
      <button type="button" @click="decrement()">+</button>
     </div>
    <script type="module">
      import Alpine from 'alpinejs';
      Alpine.start();
    </script>
  </body>
</html>
```

#### counter.js
```js
import Alpine from 'alpinejs';

Alpine.data('counter', () => ({
  count: 0,

  increment() {
     this.count++;
  },

  decrement() {
     if (this.count > 0) {
       this.count--;
     }
  },
}));
```

This works just fine of course, but the `module.esm.js` is ~100KB without any sort of compression, so it would be nice to have a minified version (e.g. `module.esm.min.js`) that could be used directly, without needing to minify it with a separate build step, or resort to a CDN that does this for us.

**Comment by ekwoka** [comment]
What is the point though?

You shouldn't be shipping `node_modules` or directly using modules like that.

You can use esm.sh

Or just bundle it yourself so you get all the other benefits of bundling.

**Reply by jonkoops** [reply]
> What is the point though?

It will make the code load faster, meaning the user will have to wait less for the page to become fully interactive. Just like the minified CDN version that already exists in the package.

> You shouldn't be shipping node_modules or directly using modules like that.

I am not, this is an example that illustrates where the code is coming from. In reality this file is being copied into a directory at build time.

> You can use esm.sh

I don't want to use a third party to deliver this kind of JavaScript out of availability and security concerns.

> Or just bundle it yourself so you get all the other benefits of bundling.

I don't want to complicate my stack with additional tooling, also there aren't really any advantages to the way my application is set up that would make bundling beneficial. There is no Node.js based tooling involved at all, the tarball is pulled from NPM at build time.

**Reply by ekwoka** [reply]
Why not pull the bundle from esm.sh at build time instead?

You can make the PR to add it if you'd like.

**Reply by jonkoops** [reply]
> Why not pull the bundle from esm.sh at build time instead?

That would still introduce a dependency on 3rd party that is not the NPM registry so I'd like to avoid that.

> You can make the PR to add it if you'd like.

Yeah, I think I will. Let me see what can be done.

**Reply by jonkoops** [reply]
Created #4182

**Reply by ekwoka** [reply]
Since I didn't mention it before, you wouldn't need any node stuff to do more bundling yourself.

esbuild is available as a binary (it's made in Go) and you can just download the tarball, and point esbuild directly at an entry file, either from a shell or FFI.
